# ISTM4212_10 Project 2
##### By Group 05: Violet(Yiqi) Chen [Group Lead], Xuening Yang, Yuyang He

>Check Working Directory

In [6]:
!pwd

/home/ubuntu/notebooks/Project2_Group05


In [7]:
#%cd ./Project2_Group05

> Load sql

In [1]:
!pip freeze | grep -E 'ipython-sql|psycopg2'

ipython-sql==0.4.0
psycopg2==2.9.3
psycopg2-binary==2.9.3


In [2]:
%load_ext sql

In [3]:
%sql postgresql://student@/week6

## First Part – ER Data Model Exercise (25 points)

#### <font color='green'>  Explanation about the graph. </font>

Before getting into the first part designing our model, we closely observed the data and try to get some sense out of them. After we identified that some data files have duplicate and some does not have a natural primary key, we decided to use RStudio to process the data. There are 3 things we did: removed some duplicates, generate some primary keys, and alter columns. </br>
+ We removed the duplicate for airports_list, B43_AIRCRAFT_INVENTORY, and CARRIER_DECODE by their primary key.
+ We generate a few primary key for those who do not have a available primary keys or composite primary keys for the easiness to use. There are REPORT_ID for ontime_reporting (not available), S_DATE for airport_weather (composite), UNIQUE_CARRIER_ID for P10_EMPLOYEE (composite).
+ We altered ontime_reporting by adding YEAR. We also altered airport_weather by selecting 7 features that are regarded as important by [raw data documentation](https://www.kaggle.com/threnjen/2019-airline-delays-and-cancellations?select=raw_data_documentation.txt), which are NAME, DATE, PRCP, SNOW, SNWD, TMAX, and AWND as well as STATION selected by us used to generate S_DATE. </br>

Further details about detail steps and R functions as well as packages we used to perform the data cleaning will be available in 2. Constructing your Database – Data Loading of Second Part.

#### <font color='bronze'>1. ER-diagram (10 points)</font>

Refer to DBMS_ER_diagram_ISTM4212.png in the folder

#### <font color='bronze'>2. Relational DB schema (10 points)</font>

Refer to Relational_DB_Schema.xlsx in the folder

#### <font color='bronze'>3. General opinion on the data dictionary files provided. Provide suggestion on improving the design of the data files. (5 points)</font>

In the data preprocessing step, we identified a few issues related to data file provided, which can be improved.<br>
1) **Duplicate**: 
i. In some data dictionary files, there are rows duplicating themselves, especially AIRPORT_COORDINATES, airports_list, and CARRIER_DECODE. They need to be cleaned to make each row unique by primary key. Also, there are duplicate column in the same data file: Entity P10_EMPLOYEES have attribute OP_UNIQUE_CARRIER and UNIQUE_CARRIER_NAME, which contain exact same information as attribute CARRIER and CARRIER_NAME under the same entity. Our suggestion is to keep only either pair given these two pairs are the same.<br>
ii. In some data file, unique key is not unique. In CARRIER_DECODE.csv, according to the data description - "Display carrier name by the airline id or carrier code" - we expect both airline and carrier code to be unique key. However, doing unique by airline id or by carrier code returns different number of rows (1686 obs. and 1744 obs. respectively). On top of that, it is reasonable that a carrier may have multiple carrier codes, but a carrier code should not be shared by multiple carrier. However, for example, OP_UNIQUE_CARRIER "4E" which belongs to British Airtours Limited is also displayed as Tanana Air Service's carrier code whose code is actually "Z3" according to our research (refers to [Bureau of Transportation Statistics](https://www.bts.gov/topics/airlines-and-airports/airline-codes)). We suspect there may be some error in data collecting process, we suggest this data file to be regenerated.
2) **Inconsistency**: We identified inconsistency in the data file, in both label/field/column names and the format of data.<br>
i. One example of different column names will be airport ID, which appear as AIRLINE_ID, ORIGIN_AIRPORT_ID. Since both AIRLINE_ID and AIRPORT_ID refers to the same thing, for the easiness of understanding and database operation, we suggest to use same attribute name (either AIRLINE_ID or AIRPORT_ID) for most entities, except for ONTIME_REPORTING which has the need to differetiate between origin and destination.<br>
ii. In the AIRPORT_WEATHER entity, the attribute DATE is formatted as both mm/dd/yyyy and yyyy-mm-dd in data file airport_weather_2019.csv but as yyyy-mm-dd in airport_weather_2020.csv. To avoid error when convert to DATE data type, we suggest to format this attribute consistently as yyyy-mm-dd.<br>
iii. Also, in the AIRPORT_WEATHER entity, the columes in 2020 and 2019 csv files are ordered different, which will create mismatch when using csvstack. Therefore, we suggest that the columns should be arranged in same order and be consistent with same amount of columns (2020 has column WT18 while 2019 does not; 2019 has columns TOBS and WT11 while 2020 does not).
3) **Difficulty of Locating Primary Key**: For some data, they are unique each row, but do not have primary keys which are easy to use. We suggest that primary key can be generate using simply index or meaningful infomation from multiple attributes to make it unique.

## Second Part – Building a Physical Data model and construct the database (75 points)

#### <font color='bronze'>1. Creating your Database Objects (20 points)</font>

##### <font color='grey'>Using the schema diagram from first part above, create the physical tables for all entities and their relationships. Make sure the tables contain all primary, foreign, and unique keys as well as any default, and check constraints (if any). The tables must also Indicate the null constraint for all attributes. (All entities from first part are required)</font>

+ Create Table WEATHER
<br/> We generate our own primary key S_DATE based on STATION and DATE. We only use 9 variable according to the raw_data_documentation.txt which says that Important features are NAME, DATE, PRCP, SNOW, SNWD, TMAX, and AWND. Further explanation will be available in the section where we copy the csv file into the table.

In [15]:
%%sql
DROP TABLE IF EXISTS WEATHER Cascade;

CREATE TABLE WEATHER (
    S_DATE varchar(50) Not NULL, 
    STATION varchar(50) Not NULL,
    NAME varchar(108) NOT NULL,
    DATE DATE NOT NULL,
    PRCP Numeric(4,2) NULL,
    SNOW Numeric(4,2) NULL,
    SNWD Numeric(4,2) NULL,
    TMAX Numeric(5,2) NULL,
    AWND Numeric(4,2) NULL,
    PRIMARY KEY (S_DATE),
    UNIQUE (STATION, DATE),
    UNIQUE (NAME, DATE)
);

COMMENT ON TABLE WEATHER is 'Airport weather by date';
COMMENT ON COLUMN WEATHER.S_DATE is 'Generated primary key based on STATION and DATE';
COMMENT ON COLUMN WEATHER.STATION is 'Location code';
COMMENT ON COLUMN WEATHER.NAME is 'Location of reading';
COMMENT ON COLUMN WEATHER.DATE is 'Date of reading';

 * postgresql://student@/week6
Done.
Done.
Done.
Done.
Done.
Done.
Done.


[]

In [16]:
%%sql
select * from weather

 * postgresql://student@/week6
0 rows affected.


[]

+ Create Table AIRPORT_COORDINATES

In [18]:
%%sql
DROP TABLE IF EXISTS AIRPORT_COORDINATES Cascade;

CREATE TABLE AIRPORT_COORDINATES (
  ORIGIN_AIRPORT_ID   Numeric(5) Not NULL, 
  DISPLAY_AIRPORT_NAME   varchar(50) NOT NULL,
  LATITUDE Numeric(10,7) NULL,
  LONGITUDE  Numeric(10,7) NULL, 
  PRIMARY KEY (ORIGIN_AIRPORT_ID),
  UNIQUE (DISPLAY_AIRPORT_NAME)
);

COMMENT ON TABLE AIRPORT_COORDINATES is
'AIRPORT_COORDINATES: Latitude and Longitude coordinates of airports, by airport name and airport id';
COMMENT ON COLUMN AIRPORT_COORDINATES.ORIGIN_AIRPORT_ID is 'Airport ID, matches to ORIGIN_AIRPORT_ID in other files';
COMMENT ON COLUMN AIRPORT_COORDINATES.DISPLAY_AIRPORT_NAME is 'Display Airport, matches to DISPLAY_AIRPORT_NAME in other files';

 * postgresql://student@/week6
Done.
Done.
Done.
Done.
Done.


[]

In [19]:
%%sql
select * from AIRPORT_COORDINATES

 * postgresql://student@/week6
0 rows affected.


[]

+ Create Table AIRPORT_LIST

In [20]:
%%sql
DROP TABLE IF EXISTS AIRPORT_LIST Cascade;

CREATE TABLE AIRPORT_LIST (
  ORIGIN_AIRPORT_ID   Numeric(5) Not NULL, 
  DISPLAY_AIRPORT_NAME   varchar(108) NOT NULL,
  ORIGIN_CITY_NAME varchar(50) NULL,
  NAME  varchar(108) NULL, 
  PRIMARY KEY (ORIGIN_AIRPORT_ID),
  UNIQUE (DISPLAY_AIRPORT_NAME)
);

COMMENT ON TABLE AIRPORT_LIST is
'AIRPORT_COORDINATES: Airport display names; used with weather files';
COMMENT ON COLUMN AIRPORT_LIST.ORIGIN_AIRPORT_ID is 'Airport ID, matches to ORIGIN_AIRPORT_ID in other files';
COMMENT ON COLUMN AIRPORT_LIST.DISPLAY_AIRPORT_NAME is 'Display Airport, matches to DISPLAY_AIRPORT_NAME in other files';
COMMENT ON COLUMN AIRPORT_LIST.NAME is 'Matches to NAME in airport_weather';

 * postgresql://student@/week6
Done.
Done.
Done.
Done.
Done.
Done.


[]

In [21]:
%%sql
select * from airport_list

 * postgresql://student@/week6
0 rows affected.


[]

+ Create Table AIRCRAFT

In [22]:
%%sql
DROP TABLE IF EXISTS AIRCRAFT Cascade;

CREATE TABLE AIRCRAFT (
  MANUFACTURE_YEAR Numeric(4) NULL, 
  TAIL_NUM   varchar(50) NOT NULL,
  NUMBER_OF_SEATS Numeric(3) NULL,
  PRIMARY KEY (TAIL_NUM)
);

COMMENT ON TABLE AIRCRAFT is
'AIRCRAFT: Manufacture year and passenger capacity for aircraft by tail number';
COMMENT ON COLUMN AIRCRAFT.TAIL_NUM is 'Unique tail number, matches to TAIL_NUM in other files';

 * postgresql://student@/week6
Done.
Done.
Done.
Done.


[]

In [23]:
%%sql
select * from aircraft

 * postgresql://student@/week6
0 rows affected.


[]

+ Create Table CARRIER
<br/> may need to redo data cleaning, do unique to carrier name

In [24]:
%%sql
DROP TABLE IF EXISTS CARRIER Cascade;

CREATE TABLE CARRIER (
  AIRLINE_ID Numeric(5) NULL, --ATTENTATIVE
  OP_UNIQUE_CARRIER varchar(50) NOT NULL,
  CARRIER_NAME varchar(108) NULL,
  PRIMARY KEY (OP_UNIQUE_CARRIER)
 );

COMMENT ON TABLE CARRIER is 'CARRIER: Display carrier name by the airline id or carrier code';
COMMENT ON COLUMN CARRIER.AIRLINE_ID is 'Airport ID, matches to AIRLINE_ID in other files';
COMMENT ON COLUMN CARRIER.OP_UNIQUE_CARRIER is 'Carrier code, matches to OP_UNIQUE_CARRIER in other files';
COMMENT ON COLUMN CARRIER.CARRIER_NAME is 'Carrier name, matches to UNIQUE_CARRIER_NAME or CARRIER_NAME in other files';

 * postgresql://student@/week6
Done.
Done.
Done.
Done.
Done.
Done.


[]

In [25]:
%%sql
select * from carrier

 * postgresql://student@/week6
0 rows affected.


[]

+ Create Table P10_EMPLOYEES </br>
We recognized that this entity has a composite primary key -- OP_UNIQUE_CARRIER and ENTITY. For the easiness of use, UNIQUE_CARRIER_ID is a generated primary key based on OP_UNIQUE_CARRIER and ENTITY

In [27]:
%%sql
DROP TABLE IF EXISTS P10_EMPLOYEES Cascade;

CREATE TABLE P10_EMPLOYEES (
    UNIQUE_CARRIER_ID varchar(50) NOT NULL,
    YEAR Numeric(4)NOT NULL,
    AIRLINE_ID Numeric(5) NOT NULL,
    OP_UNIQUE_CARRIER varchar(50) NOT NULL,
    UNIQUE_CARRIER_NAME varchar(108) NOT NULL,
    ENTITY char(1) NOT NULL,
    GENERAL_MANAGE Numeric(4) NULL,
    PILOTS_COPILOTS Numeric(4) NULL,
    OTHER_FLT_PERS Numeric(4) NULL,
    PASS_GEN_SVC_ADMIN Numeric(5) NULL,
    MAINTENANCE Numeric(5) NULL,
    ARCFT_TRAF_HANDLING_GRP1 Numeric(4) NULL,
    GEN_ARCFT_TRAF_HANDLING Numeric(4) NULL,
    AIRCRAFT_CONTROL Numeric(4) NULL,
    PASSENGER_HANDLING Numeric(5) NULL,
    CARGO_HANDLING Numeric(6) NULL,
    TRAINEES_INTRUCTOR Numeric(4) NULL,
    STATISTICAL Numeric(5) NULL,
    TRAFFIC_SOLICITERS Numeric(4) NULL,
    OTHER Numeric(5) NULL,
    TRANSPORT_RELATED Numeric(5) NULL,
    TOTAL Numeric(5) NULL,
    PRIMARY KEY (UNIQUE_CARRIER_ID),
    UNIQUE (OP_UNIQUE_CARRIER, ENTITY)
    -- FOREIGN KEY (OP_UNIQUE_CARRIER) references CARRIER(OP_UNIQUE_CARRIER)
);

COMMENT ON TABLE P10_EMPLOYEES is 'P10_EMPLOYEES: Employees per carrier';
COMMENT ON COLUMN P10_EMPLOYEES.AIRLINE_ID is 'Airport ID, matches to AIRLINE_ID in other files';
COMMENT ON COLUMN P10_EMPLOYEES.OP_UNIQUE_CARRIER is 'Carrier code, matches to OP_UNIQUE_CARRIER in other files';
COMMENT ON COLUMN P10_EMPLOYEES.UNIQUE_CARRIER_NAME is 'Carrier name, matches to UNIQUE_CARRIER_NAME or CARRIER_NAME in other files';

 * postgresql://student@/week6
Done.
Done.
Done.
Done.
Done.
Done.


[]

In [28]:
%%sql
select * from P10_EMPLOYEES

 * postgresql://student@/week6
0 rows affected.


[]

+ Create Table ONTIME_REPORTING </br>
Compared to index, we rather use a more meaningful derived primary key. REPORT_ID is a generated primary key using OP_UNIQUE_CARRIER, OP_CARRIER_FL_NUM, ORIGIN_AIRPORT_ID, YEAR, MONTH, DAY_OF_MONTH, TAIL_NUM based on the assumption that a **aircraft** of a certain **carrier** can only engage in a certain **air route** departing from certain **airport** in a certain **day**.

In [29]:
%%sql
DROP TABLE IF EXISTS ONTIME_REPORTING Cascade;

CREATE TABLE ONTIME_REPORTING (
  REPORT_ID   varchar(50) Not NULL, 
  YEAR   numeric(4) NOT NULL,
  MONTH  numeric(1) NOT NULL,
  DAY_OF_MONTH   numeric(2) NOT NULL,
  DAY_OF_WEEK    numeric(1)  NOT NULL,   
  OP_UNIQUE_CARRIER varchar(10) NOT NULL,
  TAIL_NUM  varchar(10) NULL,
  OP_CARRIER_FL_NUM numeric(4) NOT NULL,
  ORIGIN_AIRPORT_ID numeric(5) NOT NULL,
  ORIGIN  CHARACTER(10) NOT NULL,
  ORIGIN_CITY_NAME varchar(50) NOT NULL,
  DEST_AIRPORT_ID numeric(5) NOT NULL,
  DEST  varchar(15) NOT NULL,
  DEST_CITY_NAME  varchar(50) NOT NULL,
  CRS_DEP_TIME numeric(4) NOT NULL,
  DEP_TIME  numeric(4) NULL,
  DEP_DELAY_NEW  numeric(6) NULL,
  DEP_DEL15 numeric(2) NULL,
  DEP_TIME_BLK VARCHAR(10) NOT NULL,
  CRS_ARR_TIME  numeric(5) NOT NULL,
  ARR_TIME   numeric(4) NULL,
  ARR_DELAY_NEW  numeric(4) NULL,
  ARR_TIME_BLK VARCHAR(10) NOT NULL,
  CANCELLED  numeric(2) NULL,
  CANCELLATION_CODE VARCHAR(1) NULL,
  CRS_ELAPSED_TIME numeric(4) NULL,
  ACTUAL_ELAPSED_TIME numeric(4) NULL,
  DISTANCE  numeric(4) NOT NULL,
  DISTANCE_GROUP numeric(2) NOT NULL,
  CARRIER_DELAY  numeric(4) NULL,
  WEATHER_DELAY  numeric(4) NULL,
  NAS_DELAY  numeric(4) NULL,
  SECURITY_DELAY numeric(4) NULL,
  LATE_AIRCRAFT_DELAY numeric(4) NULL,
  X varchar(2), -- what is this?
  PRIMARY KEY (REPORT_ID),
  FOREIGN KEY (OP_UNIQUE_CARRIER) references CARRIER(OP_UNIQUE_CARRIER)
-- The foreign key in comments cannot connect because the data in the other table is not complete.
-- Futher explanation will be available in the bottom of Part 2 question 2.
  -- FOREIGN KEY (TAIL_NUM) references AIRCRAFT(TAIL_NUM),
  -- FOREIGN KEY (ORIGIN_AIRPORT_ID) references AIRPORT_COORDINATES(ORIGIN_AIRPORT_ID) ,
  -- FOREIGN KEY (DEST_AIRPORT_ID) references AIRPORT_COORDINATES(ORIGIN_AIRPORT_ID)
  -- FOREIGN KEY (ORIGIN_AIRPORT_ID) references AIRPORT_LIST(ORIGIN_AIRPORT_ID) ,
  -- FOREIGN KEY (DEST_AIRPORT_ID) references AIRPORT_LIST(ORIGIN_AIRPORT_ID)
-- UNIQUE (CNAME)
);

 * postgresql://student@/week6
Done.
Done.


[]

In [30]:
%%sql
select * from ONTIME_REPORTING

 * postgresql://student@/week6
0 rows affected.


[]

#### <font color='bronze'>2. Constructing your Database – Data Loading (10 points)</font>

##### <font color='grey'> Once you have created the tables and their constraints, construct (i.e., populate) the tables in bulk using the data files for year 2019 (Jan-Mar) and 2020 (Jan-Mar). Make sure to check your answer (using select queries) against the original text files (using Linux or csvkit commands) </font>

%%sql 
-- checking what table do we have
SELECT *
FROM pg_catalog.pg_tables where tableowner='student'

+ Import AIRPORT_COORDINATES

Data cleaning using **RStudio**: </br>

We remove the duplicate for our selected primary key using </br>
> df <- df[!duplicated(df$origin_airport_id),]

In [33]:
!wc -l AIRPORT_COORDINATES.csv

6487 AIRPORT_COORDINATES.csv


In [34]:
%%sql
COPY AIRPORT_COORDINATES FROM '/home/ubuntu/notebooks/Project2_Group05/AIRPORT_COORDINATES.csv'
CSV
HEADER;

 * postgresql://student@/week6
6486 rows affected.


[]

We used Linux to check the rows affected then we got 6487 rows which is one more than 6486 rows in the original dataset, AIRPORT_COORDINATES.csv. This is because we excluded the header when we inspect the data, therefore we used the same data in the next constructing table step.

In [38]:
!head -3 AIRPORT_COORDINATES.csv

"ORIGIN_AIRPORT_ID","DISPLAY_AIRPORT_NAME","LATITUDE","LONGITUDE"
10001,"Afognak Lake Airport",58.10944444,-152.9066667
10003,"Bear Creek Mining Strip",65.54805556,-161.0716667


In [35]:
%%sql
select * from AIRPORT_COORDINATES limit 2

 * postgresql://student@/week6
2 rows affected.


[(Decimal('10001'), 'Afognak Lake Airport', Decimal('58.1094444'), Decimal('-152.9066667')),
 (Decimal('10003'), 'Bear Creek Mining Strip', Decimal('65.5480556'), Decimal('-161.0716667'))]

+ Import weather </br>

Data cleaning using **RStudio**: </br>

The column names and column number of 2019 and 2020 csv file is different, which hindered us from directly using csvstack, To solve this problem, we used RStudio to select only the useful attributes and arrange them in certain order.According to the raw_data_documentation.txt which says that Important features are NAME, DATE, PRCP, SNOW, SNWD, TMAX, and AWND, we only include these attributes and STATION as well as our generated attribute S_DATE. We select using function
> df <-  select(df,c("STATION", "NAME", "DATE", "PRCP", "SNOW, SNWD, TMAX, and AWND))

We observed that there are 2 formats of date presented in Date. Thus, we convert it to the standard DATE data type used in SQL by
> library(lubridate) </br>
mdy <- mdy(df$DATE) </br>
dmy <- ymd(df$DATE) </br>
mdy[is.na(mdy)] <- dmy[is.na(mdy)] </br>
df$DATE<- mdy </br>

As we are awared of the composite unique key (station, date) or unique key (name, date), we decide to generate a primary key based on STATION and DATE </br>
> df <- mutate(df,S_DATE = paste0(STATION, DATE), .before = "STATION")

and remove the slash using </br>
> df$S_DATE <- gsub("-","",df$S_DATE)

We cut out the data from January to March for 2019 weather data using
> df <- df[df$DATE >= "2019-01-01" & df$DATE <= "2019-03-31", ]

In [5]:
!csvstack airport_weather_20*.csv > airport_weather.csv

In [36]:
!wc -l airport_weather.csv

17084 airport_weather.csv


In [40]:
%%sql
COPY WEATHER FROM '/home/ubuntu/notebooks/Project2_Group05/airport_weather.csv'
CSV
HEADER;

 * postgresql://student@/week6
17083 rows affected.


[]

We used Linux to check the rows affected then we got 17084 rows which is one more than 17083 rows in the original dataset, airport_weather.csv. This is because we excluded the header when we inspect the data, therefore we used the same data in the next constructing table step.

In [39]:
!head -3 airport_weather.csv

S_DATE,STATION,NAME,DATE,PRCP,SNOW,SNWD,TMAX,AWND
USW0001387420190101,USW00013874,"ATLANTA HARTSFIELD JACKSON INTERNATIONAL AIRPORT, GA US",2019-01-01,0.14,0,0,66,4.7
USW0001387420190102,USW00013874,"ATLANTA HARTSFIELD JACKSON INTERNATIONAL AIRPORT, GA US",2019-01-02,0.57,0,0,59,4.92


In [41]:
%%sql
select * from WEATHER limit 2

 * postgresql://student@/week6
2 rows affected.


[('USW0001387420190101', 'USW00013874', 'ATLANTA HARTSFIELD JACKSON INTERNATIONAL AIRPORT, GA US', datetime.date(2019, 1, 1), Decimal('0.14'), Decimal('0.00'), Decimal('0.00'), Decimal('66.00'), Decimal('4.70')),
 ('USW0001387420190102', 'USW00013874', 'ATLANTA HARTSFIELD JACKSON INTERNATIONAL AIRPORT, GA US', datetime.date(2019, 1, 2), Decimal('0.57'), Decimal('0.00'), Decimal('0.00'), Decimal('59.00'), Decimal('4.92'))]

+ Import AIRPORT_LIST

Data cleaning: </br>
We did not do cleaning to this file because each row is uniquely identified by ORIGIN_AIRPORT_ID.

In [42]:
!wc -l airports_list.csv

98 airports_list.csv


In [43]:
%%sql
COPY AIRPORT_LIST FROM '/home/ubuntu/notebooks/Project2_Group05/airports_list.csv'
CSV
HEADER;

 * postgresql://student@/week6
97 rows affected.


[]

We used Linux to check the rows affected then we got 98 rows which is one more than 97 rows in the original dataset, airport_list.csv. This is because we excluded the header when we inspect the data, therefore we used the same data in the next constructing table step.

In [44]:
!head -3 airports_list.csv

ORIGIN_AIRPORT_ID,DISPLAY_AIRPORT_NAME,ORIGIN_CITY_NAME,NAME
12992,Adams Field,"Little Rock, AR","NORTH LITTLE ROCK AIRPORT, AR US"
10257,Albany International,"Albany, NY","ALBANY INTERNATIONAL AIRPORT, NY US"


In [16]:
%%sql
select * from AIRPORT_LIST limit 2

 * postgresql://student@/week6
3 rows affected.


[(Decimal('12992'), 'Adams Field', 'Little Rock, AR', 'NORTH LITTLE ROCK AIRPORT, AR US'),
 (Decimal('10257'), 'Albany International', 'Albany, NY', 'ALBANY INTERNATIONAL AIRPORT, NY US'),
 (Decimal('10140'), 'Albuquerque International Sunport', 'Albuquerque, NM', 'ALBUQUERQUE INTERNATIONAL AIRPORT, NM US')]

+ Import AIRCRAFT

Data cleaning using **RStudio**: </br>
We check the duplicate based on selected primary key TAIL_NUM by
> df[duplicated(df$TAIL_NUM),]

We notice that some duplicates have complete records while the others do not. Thus, we create a table with duplicated record using
> dup <- df[duplicated(df$TAIL_NUM),] </br>
plyr::match_df(df,dup, on = 'TAIL_NUM')

and then careful remove duplicates while preserving those with non-zero NUMBER_OF_SEATS. Then, we get rid of all duplicate of the same records by 
> df <- df[!duplicated(df$TAIL_NUM),]

Besides, there is a cross in around the six-thousandth line in the data file causing error saying "the file is not UTF-8 encoded." We simply find the cross and delete it to solve the error.

In [46]:
!wc -l B43_AIRCRAFT_INVENTORY.csv

7362 B43_AIRCRAFT_INVENTORY.csv


We used Linux to check the rows affected then we got 7362 rows which is one more than 7361 rows in the original dataset, B43_AIRCRAFT_INVENTORY.csv. This is because we excluded the header when we inspect the data, therefore we used the same data in the next constructing table step.

In [47]:
%%sql
COPY AIRCRAFT FROM '/home/ubuntu/notebooks/Project2_Group05/B43_AIRCRAFT_INVENTORY.csv'
CSV
HEADER;

 * postgresql://student@/week6
7361 rows affected.


[]

In [45]:
!head -3 B43_AIRCRAFT_INVENTORY.csv

"MANUFACTURE_YEAR","TAIL_NUM","NUMBER_OF_SEATS"
1944,"N54514",0
1945,"N1651M",0


In [48]:
%%sql
select * from AIRCRAFT limit 2

 * postgresql://student@/week6
2 rows affected.


[(Decimal('1944'), 'N54514', Decimal('0')),
 (Decimal('1945'), 'N1651M', Decimal('0'))]

+ Import P10_EMPLOYEES

Data cleaning using **RStudio**: </br>
We are awared of the composite primary key (OP_UNIQUE_CARRIER, ENTITY). But for the easiness of use, we decided to generate a primary key based on OP_UNIQUE_CARRIER and ENTITY
> df <- mutate(df,UNIQUE_CARRIER_ID = paste0(OP_UNIQUE_CARRIER, ENTITY), .before = "YEAR")

In [50]:
!wc -l P10_EMPLOYEES.csv

90 P10_EMPLOYEES.csv


In [51]:
%%sql
COPY P10_EMPLOYEES FROM '/home/ubuntu/notebooks/Project2_Group05/P10_EMPLOYEES.csv'
CSV
HEADER;

 * postgresql://student@/week6
89 rows affected.


[]

We used Linux to check the rows affected then we got 90 rows which is one more than 89 rows in the original dataset, P10_EMPLOYEES.csv This is because we excluded the header when we inspect the data, therefore we used the same data in the next constructing table step.

In [52]:
!head -3 P10_EMPLOYEES.csv

"UNIQUE_CARRIER_ID","YEAR","AIRLINE_ID","OP_UNIQUE_CARRIER","UNIQUE_CARRIER_NAME","ENTITY","GENERAL_MANAGE","PILOTS_COPILOTS","OTHER_FLT_PERS","PASS_GEN_SVC_ADMIN","MAINTENANCE","ARCFT_TRAF_HANDLING_GRP1","GEN_ARCFT_TRAF_HANDLING","AIRCRAFT_CONTROL","PASSENGER_HANDLING","CARGO_HANDLING","TRAINEES_INTRUCTOR","STATISTICAL","TRAFFIC_SOLICITERS","OTHER","TRANSPORT_RELATED","TOTAL"
"0WQD",2019,21352,"0WQ","Avjet Corporation","D",4,53,6,19,29,21,0,0,0,3,1,18,0,7,0,161
"1BQI",2019,21492,"1BQ","Eastern Airlines f/k/a Dynamic Airways, LLC","I",14,50,0,41,18,21,0,0,0,0,1,13,0,3,0,161


In [53]:
%%sql
select * from P10_EMPLOYEES limit 2

 * postgresql://student@/week6
2 rows affected.


[('0WQD', Decimal('2019'), Decimal('21352'), '0WQ', 'Avjet Corporation', 'D', Decimal('4'), Decimal('53'), Decimal('6'), Decimal('19'), Decimal('29'), Decimal('21'), Decimal('0'), Decimal('0'), Decimal('0'), Decimal('3'), Decimal('1'), Decimal('18'), Decimal('0'), Decimal('7'), Decimal('0'), Decimal('161')),
 ('1BQI', Decimal('2019'), Decimal('21492'), '1BQ', 'Eastern Airlines f/k/a Dynamic Airways, LLC', 'I', Decimal('14'), Decimal('50'), Decimal('0'), Decimal('41'), Decimal('18'), Decimal('21'), Decimal('0'), Decimal('0'), Decimal('0'), Decimal('0'), Decimal('1'), Decimal('13'), Decimal('0'), Decimal('3'), Decimal('0'), Decimal('161'))]

+ Import CARRIER

Data cleaning using **RStudio**: </br>

We get rid of duplicate based of selected primary key OP_UNIQUE_CARRIER using
> df <- df[!duplicated(df$OP_UNIQUE_CARRIER),]

We notice that while some records have same AIRLINE_ID and CARRIER_NAME, their OP_UNIQUE_CARRIER could be different. We find it curious. However, out of prudence, we would rather not delete these records until further investigation on carrier decode is done.

In [54]:
!wc -l CARRIER_DECODE.csv

1745 CARRIER_DECODE.csv


In [55]:
%%sql
COPY CARRIER FROM '/home/ubuntu/notebooks/Project2_Group05/CARRIER_DECODE.csv'
CSV
HEADER;

 * postgresql://student@/week6
1744 rows affected.


[]

We used Linux to check the rows affected then we got 1745 rows which is one more than 1744 rows in the original dataset, CARRIER_DECODE.csv. This is because we excluded the header when we inspect the data, therefore we used the same data in the next constructing table step.

In [56]:
!head -3 CARRIER_DECODE.csv

"AIRLINE_ID","OP_UNIQUE_CARRIER","CARRIER_NAME"
21754,"2PQ","21 Air LLC"
20342,"Q5","40-Mile Air"


In [57]:
%%sql
select * from CARRIER limit 2

 * postgresql://student@/week6
2 rows affected.


[(Decimal('21754'), '2PQ', '21 Air LLC'),
 (Decimal('20342'), 'Q5', '40-Mile Air')]

+ Import ONTIME_REPORTING

Data cleaning using **RStudio**: </br>
We noticed that this dataset does not have a natural primary key. Instead of using index, we decided to generate one that allows us to uniquely identify each record. Firstly, we add a column YEAR to the files. </br>
>library(dplyr)
df <- mutate(df,YEAR = 2019, .before = "MONTH") for ontime reporting of 2019
df <- mutate(df,YEAR = 2020, .before = "MONTH") for ontime reporting of 2019

Then, we generated primary key using OP_UNIQUE_CARRIER, OP_CARRIER_FL_NUM, ORIGIN_AIRPORT_ID, YEAR, MONTH, DAY_OF_MONTH, TAIL_NUM based on the assumption that a **aircraft** of a certain **carrier** can only engage in a certain **air route** departing from certain **airport** in a certain **day**. </br>
> df <- mutate(df,REPORT_ID = paste0(OP_UNIQUE_CARRIER, OP_CARRIER_FL_NUM, ORIGIN_AIRPORT_ID, YEAR, MONTH, DAY_OF_MONTH, TAIL_NUM), .before = "YEAR")

which can uniquely identify each row, because it return 0 row when we test our assumption with 
> df[duplicated(df$REPORT_ID),]

In [15]:
!wc -l c_ONTIME_REPORTING_20*.csv

   583986 c_ONTIME_REPORTING_2019_01.csv
   533176 c_ONTIME_REPORTING_2019_02.csv
   632075 c_ONTIME_REPORTING_2019_03.csv
   607347 c_ONTIME_REPORTING_2020_01.csv
   574269 c_ONTIME_REPORTING_2020_02.csv
   648230 c_ONTIME_REPORTING_2020_03.csv
  3579083 total


In [16]:
#!csvcut -n c_ONTIME_REPORTING_2019_01.csv

In [16]:
!csvstack c_ONTIME_REPORTING_20*.csv > ONTIME_REPORTING_JANtoMAR.csv

In [58]:
!wc -l ONTIME_REPORTING_JANtoMAR.csv

3579078 ONTIME_REPORTING_JANtoMAR.csv


In [59]:
%%sql
COPY ONTIME_REPORTING FROM '/home/ubuntu/notebooks/Project2_Group05/ONTIME_REPORTING_JANtoMAR.csv'
CSV
HEADER;

 * postgresql://student@/week6
3579077 rows affected.


[]

We merged ontime reporting 2019 and 2020 together, 3579083 total, to a new file, ONTIME_REPORTING_JANtoMAR.csv. We used Linux to check the rows affected then we got 3579078 rows (5 rows missing from "c_ONTIME_REPORTING_20*.csv" because of 5 headers) which is one more than 3579077 rows in the previous dataset, ONTIME_REPORTING_JANtoMAR.csv. This is because we excluded the header when we inspect the data, therefore we used the same data in the next constructing table step.

In [60]:
!head -3 ONTIME_REPORTING_JANtoMAR.csv

REPORT_ID,YEAR,MONTH,DAY_OF_MONTH,DAY_OF_WEEK,OP_UNIQUE_CARRIER,TAIL_NUM,OP_CARRIER_FL_NUM,ORIGIN_AIRPORT_ID,ORIGIN,ORIGIN_CITY_NAME,DEST_AIRPORT_ID,DEST,DEST_CITY_NAME,CRS_DEP_TIME,DEP_TIME,DEP_DELAY_NEW,DEP_DEL15,DEP_TIME_BLK,CRS_ARR_TIME,ARR_TIME,ARR_DELAY_NEW,ARR_TIME_BLK,CANCELLED,CANCELLATION_CODE,CRS_ELAPSED_TIME,ACTUAL_ELAPSED_TIME,DISTANCE,DISTANCE_GROUP,CARRIER_DELAY,WEATHER_DELAY,NAS_DELAY,SECURITY_DELAY,LATE_AIRCRAFT_DELAY,X
9E328010397201916N8694A,2019,1,6,7,9E,N8694A,3280,10397,ATL,"Atlanta, GA",11150,CSG,"Columbus, GA",1645,1643,0,0,1600-1659,1732,1720,0,1700-1759,0,,47,37,83,1,,,,,,
9E328010397201917N8970D,2019,1,7,1,9E,N8970D,3280,10397,ATL,"Atlanta, GA",11150,CSG,"Columbus, GA",1645,1640,0,0,1600-1659,1732,1712,0,1700-1759,0,,47,32,83,1,,,,,,


In [61]:
%%sql
select * from ONTIME_REPORTING limit 2

 * postgresql://student@/week6
2 rows affected.


[('9E328010397201916N8694A', Decimal('2019'), Decimal('1'), Decimal('6'), Decimal('7'), '9E', 'N8694A', Decimal('3280'), Decimal('10397'), 'ATL       ', 'Atlanta, GA', Decimal('11150'), 'CSG', 'Columbus, GA', Decimal('1645'), Decimal('1643'), Decimal('0'), Decimal('0'), '1600-1659', Decimal('1732'), Decimal('1720'), Decimal('0'), '1700-1759', Decimal('0'), None, Decimal('47'), Decimal('37'), Decimal('83'), Decimal('1'), None, None, None, None, None, None),
 ('9E328010397201917N8970D', Decimal('2019'), Decimal('1'), Decimal('7'), Decimal('1'), '9E', 'N8970D', Decimal('3280'), Decimal('10397'), 'ATL       ', 'Atlanta, GA', Decimal('11150'), 'CSG', 'Columbus, GA', Decimal('1645'), Decimal('1640'), Decimal('0'), Decimal('0'), '1600-1659', Decimal('1732'), Decimal('1712'), Decimal('0'), '1700-1759', Decimal('0'), None, Decimal('47'), Decimal('32'), Decimal('83'), Decimal('1'), None, None, None, None, None, None)]

#### <font color='green'>  Further explanation about foreign key. </font>

In [6]:
%%sql
With diff as (
    Select origin_airport_id from ONTIME_REPORTING
    EXCEPT select origin_airport_id from AIRPORT_COORDINATES)
select count(*) from diff
-- There are 4 origin_airport_id present in ONTIME_REPORTING but not in AIRPORT_COORDINATES. 
    -- Therefore we cannot use foreign key to link this two tables.

 * postgresql://student@/week6
1 rows affected.


[(4,)]

In [7]:
%%sql
With diff as (
    Select dest_airport_id from ONTIME_REPORTING
    EXCEPT select origin_airport_id from AIRPORT_COORDINATES)
select count(*) from diff
-- There are 4 dest_airport_id present in ONTIME_REPORTING but not in AIRPORT_COORDINATES. 
    -- Therefore we cannot use foreign key to link this two tables.

 * postgresql://student@/week6
1 rows affected.


[(4,)]

In [5]:
%%sql
With diff as (
    Select tail_num from ONTIME_REPORTING
    EXCEPT select tail_num from AIRCRAFT)
select count(*) from diff
-- There are 383 tail_num present in ONTIME_REPORTING but not in AIRCRAFT. 
    -- Therefore we cannot use foreign key to link this two tables.

 * postgresql://student@/week6
1 rows affected.


[(383,)]

In [8]:
%%sql
With diff as (
    Select op_unique_carrier from ONTIME_REPORTING
    EXCEPT select op_unique_carrier from CARRIER)
select count(*) from diff
-- There are 0 op_unique_carrier present in ONTIME_REPORTING but not in CARRIER. 
    -- Therefore we can use foreign key to link this two tables.

 * postgresql://student@/week6
1 rows affected.


[(0,)]

In [11]:
%%sql
With diff as (
    Select op_unique_carrier from P10_EMPLOYEES
    EXCEPT select op_unique_carrier from CARRIER)
select count(*) from diff
-- There are 0 op_unique_carrier present in ONTIME_REPORTING but not in CARRIER. 
    -- Therefore we cannot use foreign key to link this two tables.

 * postgresql://student@/week6
1 rows affected.


[(0,)]

In [12]:
%%sql
With diff as (
    Select origin_airport_id from ONTIME_REPORTING
    EXCEPT select origin_airport_id from AIRPORT_LIST)
select count(*) from diff
-- There are 258 origin_airport_id present in ONTIME_REPORTING but not in AIRPORT_LIST.
    -- Therefore we cannot use foreign key to link this two tables.

 * postgresql://student@/week6
1 rows affected.


[(258,)]

In [13]:
%%sql
With diff as (
    Select dest_airport_id from ONTIME_REPORTING
    EXCEPT select origin_airport_id from AIRPORT_LIST)
select count(*) from diff
-- There are 258 dest_airport_id present in ONTIME_REPORTING but not in AIRPORT_LIST. 
    -- Therefore we cannot use foreign key to link this two tables.

 * postgresql://student@/week6
1 rows affected.


[(258,)]

In [14]:
%%sql
With diff as (
    Select name from AIRPORT_LIST
    EXCEPT select name from WEATHER)
select count(*) from diff
-- There are 1 name present in WEATHER but not in AIRPORT_LIST. 
    -- Therefore we cannot use foreign key to link this two tables. 

 * postgresql://student@/week6
1 rows affected.


[(1,)]

#### <font color='bronze'>3. Change Management (10 points)</font>

##### <font color='grey'> Your sponsor asked you to manipulate the original flight date. Write a query that combine day, month, and year of the on-time flight report file into one field. Name the new field “Flight_date”. Once created and checked you can drop the original three fields. For date conversion check https://www.postgresql.org/docs/12/functions-formatting.html.

Create Test_ONTIME_REPORTING and select few data to test the methodology

In [6]:
%%sql
DROP TABLE if exists Test_ONTIME_REPORTING Cascade;

create table Test_ONTIME_REPORTING as
SELECT * from ONTIME_REPORTING
WHERE op_unique_carrier = '9E' and day_of_week = 1;

SELECT * from Test_ONTIME_REPORTING LIMIT 2;

 * postgresql://student@/week6
Done.
20176 rows affected.
2 rows affected.


[('9E5452103972019325N8974C', Decimal('2019'), Decimal('3'), Decimal('25'), Decimal('1'), '9E', 'N8974C', Decimal('5452'), Decimal('10397'), 'ATL       ', 'Atlanta, GA', Decimal('12951'), 'LFT', 'Lafayette, LA', Decimal('1002'), Decimal('955'), Decimal('0'), Decimal('0'), '1000-1059', Decimal('1101'), Decimal('1043'), Decimal('0'), '1100-1159', Decimal('0'), None, Decimal('119'), Decimal('108'), Decimal('503'), Decimal('3'), None, None, None, None, None, None),
 ('9E5451107212019318N915XJ', Decimal('2019'), Decimal('3'), Decimal('18'), Decimal('1'), '9E', 'N915XJ', Decimal('5451'), Decimal('10721'), 'BOS       ', 'Boston, MA', Decimal('14492'), 'RDU', 'Raleigh/Durham, NC', Decimal('1422'), Decimal('1415'), Decimal('0'), Decimal('0'), '1400-1459', Decimal('1637'), Decimal('1634'), Decimal('0'), '1600-1659', Decimal('0'), None, Decimal('135'), Decimal('139'), Decimal('612'), Decimal('3'), None, None, None, None, None, None)]

In [7]:
%%sql
-- here we add a column called Flight_date in our test table
ALTER table Test_ONTIME_REPORTING 
ADD COLUMN Flight_date DATE NULL;

 * postgresql://student@/week6
Done.


[]

In [25]:
%%sql
-- here we test how can we concatenate the year, month, and day_of_month
Select Cast(year || '-' || month || '-' || day_of_month as date)
From Test_ONTIME_REPORTING
limit 2;

 * postgresql://student@/week6
2 rows affected.


[(datetime.date(2019, 1, 7),), (datetime.date(2019, 1, 7),)]

In [9]:
%%sql
-- here we update the concatenated data into Flight_date column of the test table and check the table
UPDATE Test_ONTIME_REPORTING
set Flight_date = Cast(year || '-' || month || '-' || day_of_month as date);

Select * from Test_ONTIME_REPORTING limit 2;

 * postgresql://student@/week6
20176 rows affected.


[]

In [11]:
%%sql
-- here we check if there is any rows of Flight_date column in the test table not updated
Select * from Test_ONTIME_REPORTING
WHERE Flight_date is null;

 * postgresql://student@/week6
0 rows affected.


[]

In [12]:
%%sql
-- here we delete the year, month, and day_of_month; then check the table
Alter table Test_ONTIME_REPORTING drop column year; 
Alter table Test_ONTIME_REPORTING drop column month;
Alter table Test_ONTIME_REPORTING drop column day_of_month;

Select * from Test_ONTIME_REPORTING limit 2;
-- since our test table looks good, we decide to work on the real ONTIME_REPORTING table

 * postgresql://student@/week6
Done.
Done.
Done.
2 rows affected.


[('9E3354123232019325N836AY', Decimal('1'), '9E', 'N836AY', Decimal('3354'), Decimal('12323'), 'ILM       ', 'Wilmington, NC', Decimal('10397'), 'ATL', 'Atlanta, GA', Decimal('1841'), Decimal('1834'), Decimal('0'), Decimal('0'), '1800-1859', Decimal('2019'), Decimal('123'), None, '2000-2059', Decimal('0'), None, Decimal('98'), None, Decimal('377'), Decimal('2'), None, None, None, None, None, None, datetime.date(2019, 3, 25)),
 ('9E3414129532019318N272PQ', Decimal('1'), '9E', 'N272PQ', Decimal('3414'), Decimal('12953'), 'LGA       ', 'New York, NY', Decimal('10792'), 'BUF', 'Buffalo, NY', Decimal('2150'), Decimal('2140'), Decimal('0'), Decimal('0'), '2100-2159', Decimal('2316'), Decimal('2252'), Decimal('0'), '2300-2359', Decimal('0'), None, Decimal('86'), Decimal('72'), Decimal('292'), Decimal('2'), None, None, None, None, None, None, datetime.date(2019, 3, 18))]

<font color='green'>Working on the actual ONTIME_REPORTING</font>

In [62]:
%%sql
ALTER table ONTIME_REPORTING 
ADD COLUMN Flight_date DATE NULL;

 * postgresql://student@/week6
Done.


[]

In [63]:
%%sql
Select * from ONTIME_REPORTING limit 2;

 * postgresql://student@/week6
2 rows affected.


[('9E328010397201916N8694A', Decimal('2019'), Decimal('1'), Decimal('6'), Decimal('7'), '9E', 'N8694A', Decimal('3280'), Decimal('10397'), 'ATL       ', 'Atlanta, GA', Decimal('11150'), 'CSG', 'Columbus, GA', Decimal('1645'), Decimal('1643'), Decimal('0'), Decimal('0'), '1600-1659', Decimal('1732'), Decimal('1720'), Decimal('0'), '1700-1759', Decimal('0'), None, Decimal('47'), Decimal('37'), Decimal('83'), Decimal('1'), None, None, None, None, None, None, None),
 ('9E328010397201917N8970D', Decimal('2019'), Decimal('1'), Decimal('7'), Decimal('1'), '9E', 'N8970D', Decimal('3280'), Decimal('10397'), 'ATL       ', 'Atlanta, GA', Decimal('11150'), 'CSG', 'Columbus, GA', Decimal('1645'), Decimal('1640'), Decimal('0'), Decimal('0'), '1600-1659', Decimal('1732'), Decimal('1712'), Decimal('0'), '1700-1759', Decimal('0'), None, Decimal('47'), Decimal('32'), Decimal('83'), Decimal('1'), None, None, None, None, None, None, None)]

Now we have the column flight_date added, we can start input value.

In [64]:
%%sql
UPDATE ONTIME_REPORTING
set Flight_date = Cast(year || '-' || month || '-' || day_of_month as date);

 * postgresql://student@/week6
3579077 rows affected.


[]

In [65]:
%%sql
Select * from ONTIME_REPORTING limit 2;

 * postgresql://student@/week6
2 rows affected.


[('9E3342145762019130N8970D', Decimal('2019'), Decimal('1'), Decimal('30'), Decimal('3'), '9E', 'N8970D', Decimal('3342'), Decimal('14576'), 'ROC       ', 'Rochester, NY', Decimal('12953'), 'LGA', 'New York, NY', Decimal('1436'), None, None, None, '1400-1459', Decimal('1600'), None, None, '1600-1659', Decimal('1'), 'B', Decimal('84'), None, Decimal('254'), Decimal('2'), None, None, None, None, None, None, datetime.date(2019, 1, 30)),
 ('9E3371145762019120N195PQ', Decimal('2019'), Decimal('1'), Decimal('20'), Decimal('7'), '9E', 'N195PQ', Decimal('3371'), Decimal('14576'), 'ROC       ', 'Rochester, NY', Decimal('12478'), 'JFK', 'New York, NY', Decimal('520'), None, None, None, '0001-0559', Decimal('700'), None, None, '0700-0759', Decimal('1'), 'B', Decimal('100'), None, Decimal('264'), Decimal('2'), None, None, None, None, None, None, datetime.date(2019, 1, 20))]

Now our data is filled successfully. We can drop year, month, and day_of_month columns.

In [66]:
%%sql
Alter table ONTIME_REPORTING drop column year; 
Alter table ONTIME_REPORTING drop column month;
Alter table ONTIME_REPORTING drop column day_of_month;

Select * from ONTIME_REPORTING limit 2;
-- now we successfully get rid of year, month, and day_of_month columns

 * postgresql://student@/week6
Done.
Done.
Done.
2 rows affected.
(psycopg2.ProgrammingError) can't execute an empty query
[SQL: -- now we successfully get rid of year, month, and day_of_month columns]
(Background on this error at: https://sqlalche.me/e/14/f405)


#### <font color='bronze'>4. Data analysis (20 points)</font>

##### <font color='grey'> 1) Top 10 airlines in terms of number of scheduled flights. On each of the top 10 airlines show total flights, the sum of all delays by type, and total cancellations. </font>

In [67]:
%%sql
-- AIRLINE is carrier_name
-- Sum of all delays by type (in minutes)
SELECT C.carrier_name as "airline", 
    count(*) as "total scheduled flights", 
    sum(OP.carrier_delay) as "sum of carrier delay",
    sum(OP.weather_delay) as "sum of weather delay",
    sum(OP.nas_delay) as "sum of nas delay",
    sum(OP.security_delay) as "sum of security delay",
    sum(OP.late_aircraft_delay) as "sum of late aircraft delay",
    sum(OP.cancelled) as "total cancellations"
from ONTIME_REPORTING OP, CARRIER C
WHERE C.op_unique_carrier = OP.op_unique_carrier
GROUP BY C.carrier_name
ORDER BY count(*) Desc
LIMIT 10

 * postgresql://student@/week6
10 rows affected.


[('Southwest Airlines Co.', 658265, Decimal('1509844'), Decimal('74316'), Decimal('705725'), Decimal('19991'), Decimal('1759119'), Decimal('36722')),
 ('Delta Air Lines Inc.', 467185, Decimal('1327495'), Decimal('325429'), Decimal('965721'), Decimal('3363'), Decimal('1066182'), Decimal('20174')),
 ('American Airlines Inc.', 457111, Decimal('1758590'), Decimal('181434'), Decimal('1039563'), Decimal('7618'), Decimal('1799175'), Decimal('20782')),
 ('SkyWest Airlines Inc.', 407303, Decimal('2529666'), Decimal('698177'), Decimal('1131810'), Decimal('4781'), Decimal('1634730'), Decimal('17457')),
 ('United Air Lines Inc.', 292736, Decimal('729115'), Decimal('157040'), Decimal('1098214'), Decimal('546'), Decimal('1283961'), Decimal('14297')),
 ('Midwest Airline, Inc.', 163469, Decimal('362907'), Decimal('75630'), Decimal('540814'), Decimal('1059'), Decimal('613239'), Decimal('7066')),
 ('American Eagle Airlines Inc.', 153185, Decimal('423455'), Decimal('135555'), Decimal('445035'), Decimal('1148'), Decimal('721184'), Decimal('8670')),
 ('JetBlue Airways', 146661, Decimal('806009'), Decimal('40894'), Decimal('447768'), Decimal('3124'), Decimal('844491'), Decimal('6215')),
 ('Comair Inc.', 141506, Decimal('531257'), Decimal('64036'), Decimal('245468'), Decimal('2009'), Decimal('955513'), Decimal('7076')),
 ('Endeavor Air Inc.', 131567, Decimal('387508'), Decimal('103785'), Decimal('324399'), Decimal('291'), Decimal('601609'), Decimal('4754'))]

ANS:
top 10 airlines in terms of number of scheduled flights are:
1. Southwest Airlines Co.
2. Delta Air Lines Inc.
3. American Airlines Inc.
4. SkyWest Airlines Inc.
5. United Air Lines Inc.
6. Midwest Airline, Inc.
7. American Eagle Airlines Inc.
8. JetBlue Airways
9. Comair Inc.
10. Endeavor Air Inc.

In [14]:
%%sql
SELECT distinct cancelled from ONTIME_REPORTING
-- cancelled is binary so that we can use sum() function to count total cancellations

 * postgresql://student@/week6
2 rows affected.


[(Decimal('1'),), (Decimal('0'),)]

%%sql
-- AIRLINE is carrier_name </br>
-- This code is used to count how many times of delays happen by type </br>
SELECT C.carrier_name as "airline", 
    count(*) as "total scheduled flights", 
    count(OP.carrier_delay) - sum(CASE WHEN OP.carrier_delay = 0 THEN 1 ELSE 0 END) as "total carrier delay",
    count(OP.weather_delay) - sum(CASE WHEN OP.weather_delay = 0 THEN 1 ELSE 0 END) as "total weather delay",
    count(OP.nas_delay) - sum(CASE WHEN OP.nas_delay = 0 THEN 1 ELSE 0 END) as "total nas delay",
    count(OP.security_delay) - sum(CASE WHEN OP.security_delay = 0 THEN 1 ELSE 0 END) as "total security delay",
    count(OP.late_aircraft_delay) - sum(CASE WHEN OP.late_aircraft_delay = 0 THEN 1 ELSE 0 END) as "total late aircraft delay",
    sum(OP.cancelled) as "total cancellations"
from ONTIME_REPORTING OP, CARRIER C
WHERE C.op_unique_carrier = OP.op_unique_carrier
GROUP BY C.carrier_name
ORDER BY count(*) Desc
LIMIT 10

##### <font color='grey'> 2) Top 10 busiest airport pairs in terms of total scheduled flights? Of those, indicate the airlines, total number of seats they can offer. </font>

In [68]:
%%sql
SELECT AO.DISPLAY_AIRPORT_NAME || '-' || AD.DISPLAY_AIRPORT_NAME as "busiest airport pairs", 
    C.carrier_name as "airline",
    count(*) as "total scheduled flights",
    SUM(A.NUMBER_OF_SEATS) as  "total number of seats"
FROM   ONTIME_REPORTING OP, AIRCRAFT A, CARRIER C, AIRPORT_LIST AO, AIRPORT_LIST AD
WHERE  A.tail_num = OP.tail_num 
    and C.op_unique_carrier = OP.op_unique_carrier
    and OP.origin_airport_id = AO.origin_airport_id
    and OP.dest_airport_id=AD.origin_airport_id
group by AO.DISPLAY_AIRPORT_NAME || '-' || AD.DISPLAY_AIRPORT_NAME, C.carrier_name
ORDER BY count(*) Desc
limit 10

 * postgresql://student@/week6
10 rows affected.


[('Hollywood-Burbank Midpoint-Kahului Airport', 'Hawaiian Airlines Inc.', 4680, Decimal('578478')),
 ('Kahului Airport-Hollywood-Burbank Midpoint', 'Hawaiian Airlines Inc.', 4673, Decimal('577617')),
 ('William P Hobby-Dallas Love Field', 'Southwest Airlines Co.', 3320, Decimal('487752')),
 ('Dallas Love Field-William P Hobby', 'Southwest Airlines Co.', 3300, Decimal('482940')),
 ('Hollywood-Burbank Midpoint-Lihue Airport', 'Hawaiian Airlines Inc.', 3131, Decimal('385113')),
 ('Lihue Airport-Hollywood-Burbank Midpoint', 'Hawaiian Airlines Inc.', 3123, Decimal('384129')),
 ('Hollywood-Burbank Midpoint-Keahole', 'Hawaiian Airlines Inc.', 3011, Decimal('370353')),
 ('Keahole-Hollywood-Burbank Midpoint', 'Hawaiian Airlines Inc.', 3007, Decimal('369861')),
 ('Orlando International-Atlanta Municipal', 'Delta Air Lines Inc.', 2900, Decimal('582744')),
 ('Atlanta Municipal-Orlando International', 'Delta Air Lines Inc.', 2894, Decimal('581626'))]

ANS:
The top10 busiest airport pairs in terms of total scheduled flights are:
 1. Hollywood-Burbank Midpoint-Kahului Airport Hawaiian Airlines Inc.
 2. Kahului Airport-Hollywood-Burbank Midpoint Hawaiian Airlines Inc.
 3. William P Hobby-Dallas Love Field Southwest Airlines Co.
 4. Dallas Love Field-William P Hobby Southwest Airlines Co.
 5. Hollywood-Burbank Midpoint-Lihue Airport Hawaiian Airlines Inc.
 6. Lihue Airport-Hollywood-Burbank Midpoint Hawaiian Airlines Inc.
 7. Hollywood-Burbank Midpoint-Keahole Hawaiian Airlines Inc.
 8. Keahole-Hollywood-Burbank Midpoint Hawaiian Airlines Inc.
 9. Orlando International-Atlanta Municipal Delta Air Lines Inc.
 10. Atlanta Municipal-Orlando International Delta Air Lines Inc.

##### <font color='grey'> 3) Top 10 snow days (SNOW field). For each find the corresponding date and airport. Did any flight departed or landed from those airports on those days? </font>

In [69]:
%%sql
WITH
T10SNOW_D as (
    SELECT AL.display_airport_name, W.date, W.snow, count(*) - sum(cancelled) as "flight_departed"
    FROM WEATHER W LEFT OUTER JOIN AIRPORT_LIST AL ON (LOWER(AL.name) = LOWER(W.name)), ONTIME_REPORTING OP
    WHERE W.date = OP.flight_date and OP.origin_airport_id = AL.origin_airport_id
    GROUP BY AL.display_airport_name, W.date, W.snow 
    ORDER BY W.snow Desc NULLS LAST
    LIMIT 10),
T10SNOW_L as (
    SELECT AL.display_airport_name, W.date, W.snow, count(*) - sum(cancelled) as "flight_landed"
    FROM WEATHER W LEFT OUTER JOIN AIRPORT_LIST AL ON (LOWER(AL.name) = LOWER(W.name)), ONTIME_REPORTING OP
    WHERE W.date = OP.flight_date and OP.dest_airport_id = AL.origin_airport_id
    GROUP BY AL.display_airport_name, W.date, W.snow 
    ORDER BY W.snow Desc NULLS LAST
    LIMIT 10)
SELECT D.display_airport_name, D.date, D.snow, D.flight_departed, L.flight_landed
FROM T10SNOW_D D
LEFT JOIN T10SNOW_L L ON (D.display_airport_name = L.display_airport_name) and (D.date = L.date)
ORDER BY D.snow Desc;

 * postgresql://student@/week6
10 rows affected.


[('Greater Buffalo International', datetime.date(2019, 1, 25), Decimal('17.20'), Decimal('61'), Decimal('60')),
 ('Greater Buffalo International', datetime.date(2019, 1, 30), Decimal('13.60'), Decimal('2'), Decimal('2')),
 ('Albany International', datetime.date(2019, 1, 20), Decimal('10.40'), Decimal('1'), Decimal('3')),
 ('Logan International', datetime.date(2019, 3, 4), Decimal('9.90'), Decimal('255'), Decimal('298')),
 ('Syracuse Hancock International', datetime.date(2019, 1, 20), Decimal('9.30'), Decimal('13'), Decimal('20')),
 ('Syracuse Hancock International', datetime.date(2020, 2, 7), Decimal('9.20'), Decimal('26'), Decimal('26')),
 ('Minneapolis-St Paul International', datetime.date(2019, 2, 20), Decimal('9.10'), Decimal('18'), Decimal('21')),
 ('Anchorage International', datetime.date(2020, 2, 18), Decimal('8.90'), Decimal('39'), Decimal('41')),
 ('Salt Lake City International', datetime.date(2020, 2, 3), Decimal('8.60'), Decimal('336'), Decimal('335')),
 ('General Mitchell Field', datetime.date(2019, 2, 12), Decimal('8.30'), Decimal('61'), Decimal('66'))]

ANS:
    Top 10 snow days( airport and date):
1. Greater Buffalo International     2019-01-25
2. Greater Buffalo International     2019-01-30
3. Albany International              2019-01-20
4. Logan International               2019-03-04
5. Syracuse Hancock International    2019-01-20
6. Syracuse Hancock International    2020-02-07
7. Minneapolis-St Paul International 2019-02-20
8. Anchorage International           2020-02-18
9. Salt Lake City International      2020-02-03
10.General Mitchell Field            2019-02-12

We use WITH to generate 2 tables and combine them to display aircraft departed and landed. All the airports of the top 10 snow have flight departed and landed in the certain day. The accurate number of flight departed and landed showed above in the table.

<font color='green'>Explanation</font>

We test our result by running a query on WEATHER table only. </br> 
We find out that 'BURLINGTON INTERNATIONAL AIRPORT, VT US' is not on our departed-lander-of-top-10-snow-day query while is presented on top-10-snow-day query. </br> We further investigate by looking for the 'BURLINGTON INTERNATIONAL AIRPORT, VT US' in AIRPORT_LIST table. By doing so, we find out that it is not on departed-lander-of-top-10-snow-day query because it is not available in AIRPORT_LIST AL table and thus the full name (WEATHER.name) cannot match to the ONTIME_REPORTING OP(origin_airport_id/dest_airport_id) through AIRPORT_LIST table (AL.name = WEATHER.name and OP.origin_airport_id/OP.origin_airport_id = AL.origin_airport_id). </br> 
Therefore, our query gives the corresponding top 11 snow days with the omission of 'BURLINGTON INTERNATIONAL AIRPORT, VT US' which eventually affect 10 rows.

In [60]:
%%sql
SELECT * FROM WEATHER ORDER BY snow Desc NULLS LAST LIMIT 11

 * postgresql://student@/week6
11 rows affected.


[('USW0001473320190125', 'USW00014733', 'BUFFALO NIAGARA INTERNATIONAL, NY US', datetime.date(2019, 1, 25), Decimal('0.87'), Decimal('17.20'), Decimal('5.90'), Decimal('28.00'), Decimal('18.57')),
 ('USW0001474220190120', 'USW00014742', 'BURLINGTON INTERNATIONAL AIRPORT, VT US', datetime.date(2019, 1, 20), Decimal('0.50'), Decimal('15.60'), Decimal('18.10'), Decimal('5.00'), Decimal('10.96')),
 ('USW0001473320190130', 'USW00014733', 'BUFFALO NIAGARA INTERNATIONAL, NY US', datetime.date(2019, 1, 30), Decimal('0.89'), Decimal('13.60'), Decimal('18.90'), Decimal('11.00'), Decimal('17.67')),
 ('USW0001473520190120', 'USW00014735', 'ALBANY INTERNATIONAL AIRPORT, NY US', datetime.date(2019, 1, 20), Decimal('1.25'), Decimal('10.40'), Decimal('11.00'), Decimal('21.00'), Decimal('10.96')),
 ('USW0001473920190304', 'USW00014739', 'BOSTON, MA US', datetime.date(2019, 3, 4), Decimal('0.97'), Decimal('9.90'), None, Decimal('40.00'), Decimal('13.20')),
 ('USW0001477120190120', 'USW00014771', 'SYRACUSE HANCOCK INTERNATIONAL AIRPORT, NY US', datetime.date(2019, 1, 20), Decimal('0.57'), Decimal('9.30'), Decimal('15.00'), Decimal('19.00'), Decimal('11.86')),
 ('USW0001477120200207', 'USW00014771', 'SYRACUSE HANCOCK INTERNATIONAL AIRPORT, NY US', datetime.date(2020, 2, 7), Decimal('0.90'), Decimal('9.20'), Decimal('5.10'), Decimal('33.00'), Decimal('11.18')),
 ('USW0001492220190220', 'USW00014922', 'MINNEAPOLIS ST. PAUL INTERNATIONAL AIRPORT, MN US', datetime.date(2019, 2, 20), Decimal('0.60'), Decimal('9.10'), Decimal('11.00'), Decimal('28.00'), Decimal('10.74')),
 ('USW0002645120200218', 'USW00026451', 'ANCHORAGE TED STEVENS INTERNATIONAL AIRPORT, AK US', datetime.date(2020, 2, 18), Decimal('0.88'), Decimal('8.90'), Decimal('11.80'), Decimal('44.00'), Decimal('10.51')),
 ('USW0002412720200203', 'USW00024127', 'SALT LAKE CITY INTERNATIONAL AIRPORT, UT US', datetime.date(2020, 2, 3), Decimal('0.41'), Decimal('8.60'), Decimal('5.10'), Decimal('29.00'), Decimal('16.33')),
 ('USW0001483920190212', 'USW00014839', 'MILWAUKEE MITCHELL AIRPORT, WI US', datetime.date(2019, 2, 12), Decimal('0.84'), Decimal('8.30'), Decimal('5.90'), Decimal('34.00'), Decimal('13.20'))]

In [56]:
%%sql
SELECT *
FROM AIRPORT_LIST AL
WHERE AL.name = 'BURLINGTON INTERNATIONAL AIRPORT, VT US'

 * postgresql://student@/week6
0 rows affected.


[]

Test on our output to see if our count is accurate: this test match our output which says 'Albany International' have 1 departed and 3 landed on 2019-01-20

In [21]:
%%sql
SELECT *
FROM AIRPORT_LIST AL
WHERE AL.display_airport_name = 'Albany International'

 * postgresql://student@/week6
1 rows affected.


[(Decimal('10257'), 'Albany International', 'Albany, NY', 'ALBANY INTERNATIONAL AIRPORT, NY US')]

In [23]:
%%sql
SELECT *
FROM ONTIME_REPORTING OP
WHERE (origin_airport_id = 10257 or dest_airport_id = 10257) and flight_date = '2019-01-20' and cancelled = 0

 * postgresql://student@/week6
4 rows affected.


[('DL2228103972019120N979DL', Decimal('7'), 'DL', 'N979DL', Decimal('2228'), Decimal('10397'), 'ATL       ', 'Atlanta, GA', Decimal('10257'), 'ALB', 'Albany, NY', Decimal('1952'), Decimal('2050'), Decimal('58'), Decimal('1'), '1900-1959', Decimal('2207'), Decimal('2310'), Decimal('63'), '2200-2259', Decimal('0'), None, Decimal('135'), Decimal('140'), Decimal('853'), Decimal('4'), Decimal('0'), Decimal('0'), Decimal('63'), Decimal('0'), Decimal('0'), None, datetime.date(2019, 1, 20)),
 ('AA1905110572019120N717UW', Decimal('7'), 'AA', 'N717UW', Decimal('1905'), Decimal('11057'), 'CLT       ', 'Charlotte, NC', Decimal('10257'), 'ALB', 'Albany, NY', Decimal('1630'), Decimal('1629'), Decimal('0'), Decimal('0'), '1600-1659', Decimal('1829'), None, None, '1800-1859', Decimal('0'), None, Decimal('119'), None, Decimal('646'), Decimal('3'), None, None, None, None, None, None, datetime.date(2019, 1, 20)),
 ('DL1875102572019120N912DE', Decimal('7'), 'DL', 'N912DE', Decimal('1875'), Decimal('10257'), 'ALB       ', 'Albany, NY', Decimal('10397'), 'ATL', 'Atlanta, GA', Decimal('1400'), Decimal('957'), Decimal('1197'), Decimal('1'), '1400-1459', Decimal('1706'), Decimal('1347'), Decimal('1241'), '1700-1759', Decimal('0'), None, Decimal('186'), Decimal('230'), Decimal('853'), Decimal('4'), Decimal('0'), Decimal('1198'), Decimal('43'), Decimal('0'), Decimal('0'), None, datetime.date(2019, 1, 20)),
 ('DL1875103972019120N912DE', Decimal('7'), 'DL', 'N912DE', Decimal('1875'), Decimal('10397'), 'ATL       ', 'Atlanta, GA', Decimal('10257'), 'ALB', 'Albany, NY', Decimal('1031'), Decimal('1718'), Decimal('407'), Decimal('1'), '1000-1059', Decimal('1248'), Decimal('1956'), Decimal('428'), '1200-1259', Decimal('0'), None, Decimal('137'), Decimal('158'), Decimal('853'), Decimal('4'), Decimal('0'), Decimal('0'), Decimal('428'), Decimal('0'), Decimal('0'), None, datetime.date(2019, 1, 20))]

Breaking down the code to departed flight and landed flight 

In [11]:
%%sql
-- left out join is used to display even if departed and landed is None
SELECT AL.display_airport_name, W.date, W.snow, count(*) - sum(cancelled) as "flight departed"
FROM WEATHER W LEFT OUTER JOIN AIRPORT_LIST AL ON (LOWER(AL.name) = LOWER(W.name)), ONTIME_REPORTING OP
WHERE W.date = OP.flight_date and 
    OP.origin_airport_id = AL.origin_airport_id
GROUP BY AL.display_airport_name, W.date, W.snow 
ORDER BY W.snow Desc NULLS LAST
LIMIT 10;

 * postgresql://student@/week6
10 rows affected.


[('Greater Buffalo International', datetime.date(2019, 1, 25), Decimal('17.20'), Decimal('61')),
 ('Greater Buffalo International', datetime.date(2019, 1, 30), Decimal('13.60'), Decimal('2')),
 ('Albany International', datetime.date(2019, 1, 20), Decimal('10.40'), Decimal('1')),
 ('Logan International', datetime.date(2019, 3, 4), Decimal('9.90'), Decimal('255')),
 ('Syracuse Hancock International', datetime.date(2019, 1, 20), Decimal('9.30'), Decimal('13')),
 ('Syracuse Hancock International', datetime.date(2020, 2, 7), Decimal('9.20'), Decimal('26')),
 ('Minneapolis-St Paul International', datetime.date(2019, 2, 20), Decimal('9.10'), Decimal('18')),
 ('Anchorage International', datetime.date(2020, 2, 18), Decimal('8.90'), Decimal('39')),
 ('Salt Lake City International', datetime.date(2020, 2, 3), Decimal('8.60'), Decimal('336')),
 ('General Mitchell Field', datetime.date(2019, 2, 12), Decimal('8.30'), Decimal('61'))]

In [16]:
%%sql
-- left out join is used to display even if departed and landed is None
SELECT AL.display_airport_name, W.date, W.snow, count(*) - sum(cancelled) as "flight landed"
FROM WEATHER W LEFT OUTER JOIN AIRPORT_LIST AL ON (LOWER(AL.name) = LOWER(W.name)), ONTIME_REPORTING OP
WHERE W.date = OP.flight_date and 
    OP.dest_airport_id = AL.origin_airport_id
GROUP BY AL.display_airport_name, W.date, W.snow 
ORDER BY W.snow Desc NULLS LAST
LIMIT 10;

 * postgresql://student@/week6
10 rows affected.


[('Greater Buffalo International', datetime.date(2019, 1, 25), Decimal('17.20'), Decimal('60')),
 ('Greater Buffalo International', datetime.date(2019, 1, 30), Decimal('13.60'), Decimal('2')),
 ('Albany International', datetime.date(2019, 1, 20), Decimal('10.40'), Decimal('3')),
 ('Logan International', datetime.date(2019, 3, 4), Decimal('9.90'), Decimal('298')),
 ('Syracuse Hancock International', datetime.date(2019, 1, 20), Decimal('9.30'), Decimal('20')),
 ('Syracuse Hancock International', datetime.date(2020, 2, 7), Decimal('9.20'), Decimal('26')),
 ('Minneapolis-St Paul International', datetime.date(2019, 2, 20), Decimal('9.10'), Decimal('21')),
 ('Anchorage International', datetime.date(2020, 2, 18), Decimal('8.90'), Decimal('41')),
 ('Salt Lake City International', datetime.date(2020, 2, 3), Decimal('8.60'), Decimal('335')),
 ('General Mitchell Field', datetime.date(2019, 2, 12), Decimal('8.30'), Decimal('66'))]